---
title: "Salmonella model training"
author: "Nicole Wheeler"
date: "10/6/2016"
output: html_document
---

```{r}
library(caret)
library(randomForest)
library(ggplot2)
library(ggrepel)

# set the directory you are working from
directory <- ""

load(paste(directory, "/finalmodel.Rdata", sep=""))
lookup <- read.delim(paste(directory, "/lookup.tsv", sep=""))

# Reading in the eggNOG model scores, checking to make sure that top eggNOG model hit for each protein in the orthogroup is the same
files <- list.files(paste(directory, "/search/", sep=""), pattern=".parse")

bitscores <- data.frame()

for(file in files) {
  strain <- sub(".search.parse", "", file)
  read <- read.delim(paste(directory, "/search/", file, sep=""), stringsAsFactors=F, comment.char="@", header=F)
  bitscores <- rbind(bitscores, cbind.data.frame(strain=rep(strain, nrow(read)), read))
}

# go through each row of the ortholog list and collect bitscores for each gene - FIX THIS - need to be comparing the right gene to the right model
orthlist <- read.delim(paste(directory, "/orthlist.tsv", sep=""), header=T, stringsAsFactors=F)

testdata <- data.frame()

for(i in unique(bitscores$strain)) {  # for each sample
  scores <- vector()
  for (j in lookup$Gene) { # for each key indicator gene
    if (!i %in% bitscores$strain[bitscores$V1 %in% orthlist[orthlist[,1]==j,]]) {
        scores <- c(scores, NA)
    }
    else if(bitscores$V7[bitscores$V1 %in% orthlist[orthlist[,1]==j,]&bitscores$strain==i&bitscores$V4==lookup$Model[lookup$Gene==j]][1]<0.0001) { # if the E-value looks reasonable
    scores <- c(scores, bitscores$V8[bitscores$V1 %in% orthlist[orthlist[,1]==j,]&bitscores$strain==i&bitscores$V4==lookup$Model[lookup$Gene==j]][1])
  } else {
    scores <- c(scores, NA)
  }
  }
  testdata <- rbind(testdata, scores)
}

names(testdata) <- colnames(train5)[-197]
row.names(testdata) <- unique(bitscores$strain)

# testdata[is.na(testdata)] <- 0 # use with caution, only with high-coverage genomes

# if a gene is missing from all of your samples, values will be filled in based on the median observed in our training data
for(i in 1:ncol(testdata)) {
  if(sum(is.na(testdata[,i])) == nrow(testdata)) {
    testdata[,i] <- median(train5[,names(testdata)[i]])
  }
}

# load the random forest model and score your isolates
votes <- data.frame(predict(model5, na.roughfix(testdata), type="vote"))
train_votes <- data.frame(predict(model5, train5, type="vote"))
oob_votes <- data.frame(model5$votes)

# Visualisation

combined <- rbind(votes, train_votes)
combined$Dataset <- c(rep("Testing", nrow(votes)), rep("Training", nrow(train_votes)))
combined$Dataset_jitter <- jitter(as.numeric(as.factor(combined$Dataset)), amount=0.4)
combined$Phenotype <- c(rep("Unknown", nrow(testdata)), as.character(train5$class))

ggplot(combined, aes(x=Dataset_jitter, col=Phenotype, y=Invasive, label=row.names(combined))) + geom_point() + geom_text_repel() + coord_flip() + theme_bw() + ylim(0, 1.1) + xlab("") + ylab("Invasiveness index") + geom_hline(yintercept = 0.5)
```


```{r, how does each gene look for our strains of interest?}
scores <- data.frame()

for(i in 1:10000) {
  scores <- rbind(scores, getTree(model5, i))
}
scores <- scores[scores$`split var`!=0,]
cutoffs <- vector()
for(i in 1:196) {
  cutoffs <- c(cutoffs, median(scores$`split point`[scores$`split var`==i]))
}
degradation <- vector()
for(i in colnames(train5)[-197]) {
  degradation <- c(degradation, median(train5[1:6,i], na.rm=T)>median(train5[7:15,i], na.rm=T))
}

genedata <- data.frame()
for(i in 1:length(names(testdata))) {
  if(degradation[i]==T) {
      genedata <- rbind(genedata, c(testdata[,i]<cutoffs[i], train5[,i]<cutoffs[i]))
  } else {
      genedata <- rbind(genedata, c(testdata[,i]>cutoffs[i], train5[,i]>cutoffs[i]))
  }
}
row.names(genedata) <- sub("_gpro.*", "", names(testdata))
colnames(genedata) <-c(paste(row.names(testdata), "_test", sep=""), paste(row.names(train5), "_train", sep=""))

library(ggplot2)
library(reshape2)

genedata[genedata==TRUE] <- 1
genedata[genedata==FALSE] <- 0
melted <- melt(cbind.data.frame(Gene=row.names(genedata), genedata))
melted$value[melted$value==1] <- "Invasive"
melted$value[melted$value==0] <- "Gastrointestinal"
ggplot(melted, aes(x = Gene, y = variable, fill = value)) + geom_tile() + scale_fill_manual("Prediction", values=c("black", "white")) + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
ggsave("genes.png", width=7, height=4)

genedata[genedata==1] <- "Invasive"
genedata[genedata==0] <- "Gastrointestinal"
write.csv(genedata, file="gene_predictions.csv")
```

